src/main/java/org/broadinstitute/hellbender/engine/filters/CountingReadFilter.java

```diff
@@ -0,0 +1,179 @@
+package org.broadinstitute.hellbender.engine.filters;
+
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+
+/**
+ * Wrapper/adapter for ReadFilter that counts the number of reads filtered, and provides a filter count summary.
+ *
+ * For filters with complex predicates that are composed of compound and/or operators, provides counts
+ * by level for each component predicate.
+ *
+ */
+public class CountingReadFilter implements ReadFilter {
+
+    static private final long serialVersionUID = 1L;
+
+    // Underlying ReadFilter we delegate to if we're wrapping a simple ReadFilter
+    protected final ReadFilter delegateFilter;
+
+    // Name for this filter, used to report filter count summaries
+    protected final String name;
+
+    // Number of reads filtered by this filter
+    protected long filteredCount = 0;
+
+    public CountingReadFilter(final String filterName, final ReadFilter readFilter) {
+        Utils.nonNull(readFilter);
+        name = filterName;
+        delegateFilter = readFilter;
+    }
+
+    // Used by the private CountingBinopReadFilter subclass
+    private CountingReadFilter(final String filterName) {
+        name = filterName;
+        delegateFilter = null;
```

Comment:
Is it safe for this constructor to set `delegateFilter` to null given that `test()` invokes `delegateFilter.test()` without a null check?